lib/HotModuleReplacementPlugin.js

```diff
@@ -723,6 +723,30 @@ class HotModuleReplacementPlugin {
 										newDependentHashModules
 									);
 								}
+
+								let chunkLoadingValue;
+								if (currentChunk) {
+									const entryOptions = currentChunk.getEntryOptions();
+									chunkLoadingValue =
+										(entryOptions && entryOptions.chunkLoading) ||
+										compilation.outputOptions.chunkLoading;
+								} else {
+									chunkLoadingValue = compilation.outputOptions.chunkLoading;
+								}
+
+								// add runtime requirements for ESM format
+								if (chunkLoadingValue === "import") {
+									chunkGraph.addChunkRuntimeRequirements(
+										hotUpdateChunk,
+										new Set([
+											RuntimeGlobals.publicPath,
+											RuntimeGlobals.getChunkUpdateScriptFilename,
+											RuntimeGlobals.moduleFactoriesAddOnly,
+											RuntimeGlobals.hasOwnProperty
+										])
+									);
+								}
```

Comment:
Do we really need this here? I think we should add them in `ModuleChunkLoadingPlugin.js`. Can you provide more information here?

Reply:
@alexander-akait You're right - I've moved the runtime requirements to `ModuleChunkLoadingPlugin` where they belong.
These requirements are needed because the ESM HMR update handler uses `publicPath` to construct the import URL and `getChunkUpdateScriptFilename` to determine the update chunk filename. I noticed `JsonpChunkLoadingPlugin` follows the same pattern in its `hmrDownloadUpdateHandlers` hook.

Moved them from `HotModuleReplacementPlugin` to keep the separation clean - HMR plugin handles the update flow, chunk loading plugin handles how chunks are loaded.